plugins/glustershd/rest.go

```diff
@@ -1,197 +1,129 @@
-	if err != nil {
-		if err == errors.ErrVolNotFound {
-			restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+	lock, unlock := transaction.CreateLockFuncs(volname)
+	// Taking a lock outside the txn as volinfo.Nodes() must also
+	// be populated holding the lock.
+	if err := lock(ctx); err != nil {
+		if err == transaction.ErrLockTimeout {
+			restutils.SendHTTPError(ctx, w, http.StatusConflict, err)
 		} else {
 			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
 		}
 		return
 	}
-	// Store initial volinfo before changing the HealFlag
-	tmp := *v
-	oldvolinfo := &tmp
-
-	// validate volume type
-	if !isVolReplicate(v.Type) {
-		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume Type not supported")
-		return
-	}
-
-	// Transaction which checks if all replicate volumes are stopped before
-	// stopping the self-heal daemon.
-	txn := transaction.NewTxn(ctx)
-	defer txn.Cleanup()
+	defer unlock(ctx)
 
-	// Lock on volume name.
-	lock, unlock, err := transaction.CreateLockSteps(volname)
+	// Validate volume existence
+	volinfo, err := volume.GetVolume(volname)
 	if err != nil {
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
-		return
+		if err == gderrors.ErrVolNotFound {
+			logger.WithError(err).WithField(
+				"volname", volname).Debug("volume not found")
+			restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
+		} else {
+			logger.WithError(err).WithField(
+				"volname", "volname").Debug("error occured while looking for volume")
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		}
 	}
 
-	v.HealEnabled = false
-
-	txn.Nodes = v.Nodes()
-	txn.Steps = []*transaction.Step{
-		lock,
-		{
-			DoFunc:   "vol-option.UpdateVolinfo",
-			Nodes:    []uuid.UUID{gdctx.MyUUID},
-			UndoFunc: "selfheald-undo",
-		},
-
-		{
-			DoFunc: "selfheal-stop",
-			Nodes:  txn.Nodes,
-		},
-		{
-			DoFunc: "vol-option.NotifyVolfileChange",
-			Nodes:  txn.Nodes,
-		},
-		unlock,
+	// Validate volume type
+	if !isVolReplicate(volinfo.Type) {
+		logger.WithError(err).WithField(
+			"volname", volname).Debug("volume type is not supported")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "volume type is not supported")
 	}
 
-	if err := txn.Ctx.Set("volinfo", v); err != nil {
-		logger.WithError(err).Error("failed to set volinfo in transaction context")
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
-		return
+	// Validate volume state
+	if volinfo.State != volume.VolStarted {
+		logger.WithField("volname", volname).Debug("volume should be in started state")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, gderrors.ErrVolNotStarted)
 	}
 
-	if err := txn.Ctx.Set("oldvolinfo", oldvolinfo); err != nil {
-		logger.WithError(err).Error("failed to set volinfo in transaction context")
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
-		return
+	var options []string
```

Comment:
Move this processing to inside of `getHealInfo` function.

Make `getHealInfo` return glustershdapi.HealInfo.